From 89ab4535c76f5ec69e93e7d9aa40923654da0b86 Mon Sep 17 00:00:00 2001
From: Rokis <ghost1988102@gmail.com>
Date: Fri, 2 Apr 2021 20:40:31 +0800
Subject: [PATCH 113/122] rc: in order to display wired AiMesh node status
 correctly, replace GetPhyStatus function with my own code(major change).

Signed-off-by: Rokis <ghost1988102@gmail.com>
---
 release/src/router/httpd/web.c                |  4 --
 release/src/router/rc/Makefile                |  3 --
 release/src/router/rc/ate.c                   |  7 ---
 release/src/router/rc/init.c                  |  4 --
 release/src/router/rc/interface.c             |  4 +-
 release/src/router/rc/k3.c                    | 43 ++++++++++++++++---
 release/src/router/rc/k3.h                    |  1 -
 release/src/router/rc/rc.c                    | 16 -------
 release/src/router/rc/rc.h                    |  3 ++
 release/src/router/rc/sysdeps/init-broadcom.c |  3 +-
 release/src/router/rc/watchdog.c              |  8 ----
 .../router/shared/sysdeps/broadcom/broadcom.c | 15 ++++++-
 12 files changed, 58 insertions(+), 53 deletions(-)

diff --git a/release/src/router/httpd/web.c b/release/src/router/httpd/web.c
index 8dad9f00df..9e411c646e 100644
--- a/release/src/router/httpd/web.c
+++ b/release/src/router/httpd/web.c
@@ -28392,11 +28392,7 @@ ej_get_wan_lan_status(int eid, webs_t wp, int argc, char **argv)
 	struct json_object *wanLanLinkSpeed = NULL;
 	struct json_object *wanLanCount = NULL;
 
-#ifdef RTK3
-	fp = popen("rc Get_PhyStatus", "r");
-#else
 	fp = popen("ATE Get_WanLanStatus", "r");
-#endif
 	if (fp == NULL)
 		goto error;
 
diff --git a/release/src/router/rc/Makefile b/release/src/router/rc/Makefile
index d33e55b5d8..d30ced93d8 100644
--- a/release/src/router/rc/Makefile
+++ b/release/src/router/rc/Makefile
@@ -1205,9 +1205,6 @@ endif
 	@cd $(INSTALLDIR)/sbin && ln -sf rc get_phy_status
 	@cd $(INSTALLDIR)/sbin && ln -sf rc get_phy_speed
 	@cd $(INSTALLDIR)/sbin && ln -sf rc GetPhyStatus
-ifeq ($(RTK3),y)
-	@cd $(INSTALLDIR)/sbin && ln -sf rc Get_PhyStatus
-endif
 ifeq ($(HND_ROUTER),y)
 	@cd $(INSTALLDIR)/sbin && ln -sf rc set_ex53134_ctrl
 	@cd $(INSTALLDIR)/sbin && ln -sf rc set_phy_ctrl
diff --git a/release/src/router/rc/ate.c b/release/src/router/rc/ate.c
index 9fa057e07f..3a40d9befb 100644
--- a/release/src/router/rc/ate.c
+++ b/release/src/router/rc/ate.c
@@ -18,9 +18,6 @@
 #ifdef RTCONFIG_QCA_PLC_UTILS
 #include <plc_utils.h>
 #endif
-#ifdef RTK3
-#include "k3.h"
-#endif
 
 #define MULTICAST_BIT  0x0001
 #define UNIQUE_OUI_BIT 0x0002
@@ -2351,12 +2348,8 @@ int asus_ate_command(const char *command, const char *value, const char *value2)
 	else if (!strcmp(command, "Get_WanLanStatus")) {
 #if defined(RTCONFIG_EXT_RTL8365MB) || defined(RTCONFIG_EXT_RTL8370MB)
 		GetPhyStatus(1, NULL);
-#else
-#ifdef RTK3
-		if (!GetPhyStatusk3(1) && nvram_match("ATEMODE", "1")) {
 #else
 		if (!GetPhyStatus(1, NULL) && nvram_match(ATE_FACTORY_MODE_STR(), "1")) {
-#endif
 			puts("ATE_ERROR");
 		}
 #endif
diff --git a/release/src/router/rc/init.c b/release/src/router/rc/init.c
index e45232a6fc..d3674076fb 100644
--- a/release/src/router/rc/init.c
+++ b/release/src/router/rc/init.c
@@ -75,10 +75,6 @@
 #include <sys/statfs.h>
 #endif
 
-#ifdef RTK3
-#include "k3.h"
-#endif
-
 #define SHELL "/bin/sh"
 #define LOGIN "/bin/login"
 
diff --git a/release/src/router/rc/interface.c b/release/src/router/rc/interface.c
index 2698913fd6..1f43071eed 100644
--- a/release/src/router/rc/interface.c
+++ b/release/src/router/rc/interface.c
@@ -103,7 +103,7 @@ void _switch_gen_config(char *buf, const int ports[SWPORT_COUNT], int swmask, ch
 
 	if (wan && !cputag) {
             for (n = i = count = 0; i < SWPORT_COUNT && mask; mask >>= 1, i++) {
-                if ((mask & 1U) == 0)
+                if ((mask & 1U) == 0 || ports[i] < 0)
                         continue;
                 res[n].port = ports[i];
                 res[n].tag = (i == SWPORT_CPU) ? cputag : "";
@@ -113,7 +113,7 @@ void _switch_gen_config(char *buf, const int ports[SWPORT_COUNT], int swmask, ch
 	}
 	else {
 	    for (i = count = 0; i < SWPORT_COUNT && mask; mask >>= 1, i++) {
-		if ((mask & 1U) == 0)
+		if ((mask & 1U) == 0 || ports[i] < 0)
 			continue;
 		for (n = count; n > 0 && ports[i] < res[n - 1].port; n--)
 			res[n] = res[n - 1];
diff --git a/release/src/router/rc/k3.c b/release/src/router/rc/k3.c
index e87e339382..41f03ce6e7 100644
--- a/release/src/router/rc/k3.c
+++ b/release/src/router/rc/k3.c
@@ -25,7 +25,6 @@
 #include <siutils.h>
 #include <auth_common.h>
 #include <sys/reboot.h>
-#include "k3.h"
 
 #define ROMCFE "/rom/cfe"
 
@@ -210,24 +209,49 @@ void start_k3screen(void)
 	kprintf("k3screen: ok\n");
 }
 
-int GetPhyStatusk3(int verbose)
+int GetPhyStatus(int verbose, phy_info_list *list)
 {
 	int port[] = { 3, 1, 0, 2 };
-	int i, ret, lret = 0, mask;
-	char out_buf[20];
+	int i, ret, lret = 0, mask, ret_code = 0;
+	char out_buf[21];
 
 	bzero(out_buf, sizeof(out_buf));
 	for (i = 0; i < 4; i++)
 	{
 		mask = 0;
 		mask |= 0x0001 << port[i];
+		if (list)
+		{
+			list->count++;
+			list->phy_info[i].phy_port_id = port[i];
+			if (i == 0)
+			{
+				snprintf(list->phy_info[i].label_name, sizeof(list->phy_info[i].label_name), "W0");
+				snprintf(list->phy_info[i].cap_name, sizeof(list->phy_info[i].cap_name), "wan");
+			}
+			else
+			{
+				snprintf(list->phy_info[i].label_name, sizeof(list->phy_info[i].label_name), "L%d", i);
+				snprintf(list->phy_info[i].cap_name, sizeof(list->phy_info[i].cap_name), "lan");
+			}
+			list->phy_info[i].tx_packets = get_phy_mib(port[i], "tx_packets");
+			list->phy_info[i].rx_packets = get_phy_mib(port[i], "rx_packets");
+			list->phy_info[i].tx_bytes = get_phy_mib(port[i], "tx_bytes");
+			list->phy_info[i].rx_bytes = get_phy_mib(port[i], "rx_bytes");
+			list->phy_info[i].crc_errors = get_phy_mib(port[i], "crc_errors");
+		}
 		if (get_phy_status(mask) == 0)
 		{ /*Disconnect*/
 			if (i == 0)
 				snprintf(out_buf, sizeof(out_buf), "W0=X;");
 			else
-			{
 				snprintf(out_buf, sizeof(out_buf), "%sL%d=X;", out_buf, i);
+
+			if (list)
+			{
+				snprintf(list->phy_info[i].state, sizeof(list->phy_info[i].state), "down");
+				snprintf(list->phy_info[i].duplex, sizeof(list->phy_info[i].duplex), "none");
+				list->phy_info[i].link_rate = 0;
 			}
 		}
 		else
@@ -241,13 +265,22 @@ int GetPhyStatusk3(int verbose)
 			else
 			{
 				lret = 1;
+				ret_code |= 0x0001 << i;
 				snprintf(out_buf, sizeof(out_buf), "%sL%d=%s;", out_buf, i, (ret & 2) ? "G" : "M");
 			}
+			if (list)
+			{
+				snprintf(list->phy_info[i].state, sizeof(list->phy_info[i].state), "up");
+				snprintf(list->phy_info[i].duplex, sizeof(list->phy_info[i].duplex), (get_phy_duplex(1 << port[i])) ? "full" : "half");
+				list->phy_info[i].link_rate = (ret & 2) ? 1000 : 100;
+			}
 		}
 	}
 
 	if (verbose)
 		puts(out_buf);
+	if (verbose != 53134 && verbose != 8365)
+		lret = ret_code;
 
 	return lret;
 }
diff --git a/release/src/router/rc/k3.h b/release/src/router/rc/k3.h
index 8617c807a8..fc4f93cea7 100644
--- a/release/src/router/rc/k3.h
+++ b/release/src/router/rc/k3.h
@@ -22,4 +22,3 @@ extern void update_cfe_k3(void);
 extern void k3_init(void);
 extern void k3_init_done(void);
 extern void start_k3screen(void);
-extern int GetPhyStatusk3(int verbose);
diff --git a/release/src/router/rc/rc.c b/release/src/router/rc/rc.c
index dd9095e0fd..5e73742652 100644
--- a/release/src/router/rc/rc.c
+++ b/release/src/router/rc/rc.c
@@ -34,10 +34,6 @@
 #include <sys/reboot.h>
 #endif
 
-#ifdef RTK3
-#include "k3.h"
-#endif
-
 #ifndef ARRAYSIZE
 #define ARRAYSIZE(a) (sizeof(a) / sizeof(a[0]))
 #endif /* ARRAYSIZE */
@@ -292,24 +288,12 @@ static int rctest_main(int argc, char *argv[])
 		setup_passwd();
 	}
 #endif
-#ifdef RTK3
-	else if (strcmp(argv[1], "GetPhyStatus")==0) {
-		printf("Get Phy status:%d\n", GetPhyStatusk3(0));
-	}
-	else if (strcmp(argv[1], "Get_PhyStatus")==0) {
-		GetPhyStatusk3(1);
-	}
-	else if (strcmp(argv[1], "GetExtPhyStatus")==0) {
-		printf("Get Ext Phy status:%d\n", GetPhyStatusk3(atoi(argv[2])));
-	}
-#else
 	else if (strcmp(argv[1], "GetPhyStatus")==0) {
 		printf("Get Phy status:%d\n", GetPhyStatus(0, NULL));
 	}
 	else if (strcmp(argv[1], "GetExtPhyStatus")==0) {
 		printf("Get Ext Phy status:%d\n", GetPhyStatus(atoi(argv[2]), NULL));
 	}
-#endif
 #ifdef CONFIG_BCMWL5
 #if defined(RTCONFIG_AMAS) && defined(RTCONFIG_BHCOST_OPT)
 	else if (strcmp(argv[1], "linkrate")==0) {
diff --git a/release/src/router/rc/rc.h b/release/src/router/rc/rc.h
index 7565976dd2..3e04a0ae0a 100644
--- a/release/src/router/rc/rc.h
+++ b/release/src/router/rc/rc.h
@@ -38,6 +38,9 @@
 
 #include "sysdeps.h"
 #include <linux/version.h>
+#ifdef RTK3
+#include <k3.h>
+#endif
 #ifdef HND_ROUTER
 #include <cms_image.h>
 #include <bcm_imgif.h>
diff --git a/release/src/router/rc/sysdeps/init-broadcom.c b/release/src/router/rc/sysdeps/init-broadcom.c
index ae1896effe..58491e30c2 100644
--- a/release/src/router/rc/sysdeps/init-broadcom.c
+++ b/release/src/router/rc/sysdeps/init-broadcom.c
@@ -14,7 +14,6 @@
 #include "shared.h"
 #include "version.h"
 #include "interface.h"
-#include "k3.h"
 
 #include <sched.h>
 #include <termios.h>
@@ -1397,7 +1396,7 @@ void generate_switch_para(void)
 #else	// RTCONFIG_EXT_RTL8365MB
 			/* WAN L1 L2 L3 L4 CPU */	/*vision: WAN L1 L2 L3 L4 */
 #ifdef RTK3
-			int ports[SWPORT_COUNT] = { 3, 1, 0, 2, 4, 5 };
+			int ports[SWPORT_COUNT] = { 3, 1, 0, 2, -1, 5 };
 #else
 			int ports[SWPORT_COUNT] = { 4, 3, 2, 1, 0, 5 };
 #endif
diff --git a/release/src/router/rc/watchdog.c b/release/src/router/rc/watchdog.c
index 61b9deb903..0ae19b865a 100644
--- a/release/src/router/rc/watchdog.c
+++ b/release/src/router/rc/watchdog.c
@@ -104,10 +104,6 @@
 #endif
 #include <json.h>
 
-#ifdef RTK3
-#include "k3.h"
-#endif
-
 #define BCM47XX_SOFTWARE_RESET	0x40		/* GPIO 6 */
 #define RESET_WAIT		2		/* seconds */
 #define RESET_WAIT_COUNT	RESET_WAIT * 10 /* 10 times a second */
@@ -5110,12 +5106,8 @@ void fake_etlan_led(void)
 #if defined(RTAX86U) || defined(RTAX5700)
 	phystatus = hnd_get_phy_status("eth5");
 	if (!phystatus)
-#else
-#ifdef RTK3
-	phystatus = GetPhyStatusk3(0);
 #else
 	phystatus = GetPhyStatus(0, NULL);
-#endif
 #if defined(RTCONFIG_EXTPHY_BCM84880)
 	if ((nvram_get_int("wans_extwan") && !(phystatus & 0x3e)) || // configure 2.5G port as WAN, need to consider 1G WAN connectivity
 			(!nvram_get_int("wans_extwan") && !(phystatus & 0x1e)))  // configure 2.5G port as LAN, ignore 2.5G port
diff --git a/release/src/router/shared/sysdeps/broadcom/broadcom.c b/release/src/router/shared/sysdeps/broadcom/broadcom.c
index e3345af320..cb7446c9a1 100644
--- a/release/src/router/shared/sysdeps/broadcom/broadcom.c
+++ b/release/src/router/shared/sysdeps/broadcom/broadcom.c
@@ -878,6 +878,9 @@ get_uplinkports_linkrate(char *ifname)
 	} else {
 		lan_ports = 0;
 	}
+#endif
+#if defined(RTK3)
+		lan_ports = 3;
 #endif
 	int ports[lan_ports+1];
 	int lrate[lan_ports+1];
@@ -934,7 +937,11 @@ get_uplinkports_linkrate(char *ifname)
 	case MODEL_RTAC88U:
 	case MODEL_RTAC3100:
 		/* WAN L1 L2 L3 L4 */
+#ifdef RTK3
+		ports[0]=3; ports[1]=1; ports[2]=0; ports[3]=2;
+#else
 		ports[0]=4; ports[1]=3; ports[2]=2; ports[3]=1; ports[4]=0;
+#endif
 #if defined(RTCONFIG_EXT_RTL8365MB) || defined(RTCONFIG_EXT_RTL8370MB)
 		ext = 1;
 #endif
@@ -1591,7 +1598,7 @@ phy_port_mapping get_phy_port_mapping(void)
 		.port[6] = { .phy_port_id = S_RTL8365MB+1, .label_name = "L6", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" },
 		.port[7] = { .phy_port_id = S_RTL8365MB+2, .label_name = "L7", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" },
 		.port[8] = { .phy_port_id = S_RTL8365MB+3, .label_name = "L8", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" }
-#elif RTCONFIG_EXT_RTL8370MB
+#elif defined(RTCONFIG_EXT_RTL8370MB)
 		.count = 9,
 		.port[0] = { .phy_port_id = 0, .label_name = "W0", .cap = PHY_PORT_CAP_WAN, .max_rate = 1000, .ifname = "eth0" },
 		.port[1] = { .phy_port_id = S_RTL8365MB+2, .label_name = "L1", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" },
@@ -1602,6 +1609,12 @@ phy_port_mapping get_phy_port_mapping(void)
 		.port[6] = { .phy_port_id = S_RTL8365MB+5, .label_name = "L6", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" },
 		.port[7] = { .phy_port_id = S_RTL8365MB+6, .label_name = "L7", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" },
 		.port[8] = { .phy_port_id = S_RTL8365MB+7, .label_name = "L8", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" }
+#elif defined(RTK3)
+		.count = 4,
+		.port[0] = { .phy_port_id = 3, .label_name = "W0", .cap = PHY_PORT_CAP_WAN, .max_rate = 1000, .ifname = "eth0" },
+		.port[1] = { .phy_port_id = 1, .label_name = "L1", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" },
+		.port[2] = { .phy_port_id = 0, .label_name = "L2", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" },
+		.port[3] = { .phy_port_id = 2, .label_name = "L3", .cap = PHY_PORT_CAP_LAN, .max_rate = 1000, .ifname = "vlan1" },
 #else
 		.count = 5,
 		.port[0] = { .phy_port_id = 4, .label_name = "W0", .cap = PHY_PORT_CAP_WAN, .max_rate = 1000, .ifname = "eth0" },
-- 
2.25.1

